### PR TITLE
Add GRANT query support for hive connector

### PR DESCRIPTION
**Pull request by amrutagokhale**
This is a prototype implementation of adding parsing support for GRANT query. I am listing the issues in the code that I am aware of:
- The current parsing code doesn't reflect the exact syntax of GRANT query that we want to support. I haven't iterated over the syntax, since this implementation is a prototype. I will refine it in due course. Here are some differences between the current syntax and the sql-99 standard syntax that we want to support:
  -- Current syntax can accept only a single privilege at a time instead of a list of privileges
  -- Current syntax doesn't take the keyword ALL which implies all privileges
- There are some TODO's in the code. These are issues to keep in mind; they may need to be fixed.
- I haven't removed all the comments from the code, specially in the hive connector code. Will do so eventually.
- The current implementation of function checkCanGrantTablePrivilege() performs the check of whether the current user can grant the privilege by checking whether the current user is the owner of the table and if so, the check passes. If not, the check fails. We may have to change this as per the security policy we decide to adopt (for ex., one of the privileges a user can have will be a 'GRANT' privilege that denotes whether a user can grant privileges to other users. We will then check whether the user has the GRANT privilege instead).

### CODE REVIEW

**Comment by amrutagokhale** [comment]
@cawallin @rschlussel: can you please take a look?

**Comment by cawallin** [comment]
Re your last bullet point: it should grant table privileges if, in the underlying data store, the user has the ability to grant privileges on a table.  Maybe I'm understanding it wrong, but in your implementation it might allow a user to do things it's not allowed to do.

**Comment by amrutagokhale** [comment]
@cawallin Actually, the current implementation is more restrictive, because it only allows the owner of a table to grant privileges on that table to others. This is totally permissible, since in Hive, the owner does get the privilege to grant privileges, as noted here: https://cwiki.apache.org/confluence/display/Hive/SQL+Standard+Based+Hive+Authorization#SQLStandardBasedHiveAuthorization-Privileges -- "In the case of tables and views, the owner gets all the privileges with grant option."

**Comment by cawallin** [comment]
@amrutagokhale But that's just for Hive, that's not an assumption you can make with databases in general. I bet that in PostgreSQL/MySQL you can revoke grant access to the owner of a table.

**Comment by amrutagokhale** [comment]
@cawallin Point taken. The function checkCanGrantTablePrivilege() is only in the Hive connector code, as of now. Will keep this in mind, when I add code for other connectors.

**Comment by rschlussel-zz** [comment]
I think even for hive we should check that the user has the grant privilege.  Both because we don't want to be more restrictive than hive.  But also because the owner having grant privileges is sort of incidental, and so we don't want the implementation to rely on that.

**Comment by amrutagokhale** [comment]
Ok, will change this then in hive code too.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java [hunk 1]
```diff
@@ -886,6 +890,26 @@ private void invalidatePartitionCache(String databaseName, String tableName)
...
+        try (HiveMetastoreClient metastoreClient = clientProvider.createMetastoreClient()) {
+            HiveObjectPrivilege privilegeObject = new HiveObjectPrivilege(
+                    new HiveObjectRef(HiveObjectType.TABLE, databaseName, tableName, null, null),
+                    identity.getUser(),
+                    PrincipalType.USER, // TODO: check whether it is USER or ROLE.
+                    privilegeGrantInfo);
+            ImmutableList.Builder<HiveObjectPrivilege> privilegeBagList = ImmutableList.builder();
+            privilegeBagList.add(privilegeObject);
+
+            // Should you check whether the user already has the given privilege and if yes, return.
```
Comment (rschlussel-zz) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: Is this a todo comment?  Also, is checking privileges that much faster than granting them?

File: presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java [hunk 2]
```diff
@@ -177,6 +179,14 @@ public void checkCanSetCatalogSessionProperty(Identity identity, String property
         }
     }
 
+    @Override
+    public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege, SchemaTableName tableName)
+    {
+        if (!checkTablePermission(identity, tableName, OWNERSHIP)) {
```
Comment (rschlussel-zz) [inline presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java]: Other users can also have grant privileges if they were granted permission 'WITH GRANT OPTION'

Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java]: yes, you're right. I am going to change OWNERSHIP -> GRANT

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java (hunk 1 above)
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: It's a todo. I am not entirely sure whether it is worth checking the privileges. I added the comment, just to keep that in mind.

File: presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java (hunk 2 above)
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java]: done (in the new commits).

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java (hunk 1 above)
Comment (amrutagokhale) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: A faster way to check privileges would be to check the userTablePrivileges cache which stores the privileges. Before we do that, we may need to verify that the cache is populated, flushed and refreshed correctly. But this is definitely something to keep in mind.

**Comment by amrutagokhale** [comment]
@petroav @rschlussel can you please take a look at the new commits? I added 3 of them. From oldest to newest, the commits are:
1) Change the GRANT syntax to be able to specify multiple privileges in a single line. For ex., "grant insert, select to user on table_name". These changes are in the parser module.
2) Changes to the hive metastore to take into account the cache used to store user privileges. Main change in CachingHiveMetastore.java file is to return an object of type "RetryDriver" (an instance of RetryDriver class runs the code for a certain maximum number of attempts and times out after that)
3) Addresses the comments. This includes changing and renaming two classes:
PrestoIdentity -> IdentityNode and PrestoPrivilege -> PrivilegeNode. I also had to add a dependency in presto-parser module on presto-spi module, since the two classes in parser module now have members from the classes in SPI interface.

File: presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        Set<Privilege> privileges = new HashSet<>();
+
+        for (PrivilegeNode privilegeNode : statement.getPrivilegeNodes()) {
+            accessControl.checkCanGrantTablePrivilege(session.getIdentity(), privilegeNode.getPrivilege(), tableName);
+
+            // TODO: if [WITH GRANT OPTION] is present, check that the current user has the ("GRANT") privilege to grant the privilege
+
+            privileges.add(privilegeNode.getPrivilege());
+        }
+        // I don't fully understand the Identity object; what is user and what is principal? For now, I am passing grantee
```
Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java]: I think this should be investigated more to ensure that it's done right.

**Comment by rschlussel-zz** [comment]
The rest of the code looks good.  What's the plan for addressing the TODOs?

File: presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java (hunk 3 above)
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java]: The "user" field corresponds to the user name and the "principal" field is the kerberos principal. @petroav pointed this out earlier. So passing the grantee's name as the first argument should be correct. With my new refactoring, this code to create a new identity has now moved to sql/tree/AstBuilder.java.

Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java]: Can you just remove the comment then.

**Comment by amrutagokhale** [comment]
@rschlussel All the TODOs are optimizations which may not be necessary to add right away. My plan is to leave the code as it is. I will clean up some of the old TODOs.

File: presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java (hunk 3 above)
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/execution/GrantTask.java]: Yup, will do.

**Comment by petroav** [comment]
I'm done reviewing. Have a look at my comments and let me know what you think. Before putting out a PR against FB Presto, I would re-organize the commits and give them better commit messages. Specifically the last refactor commit (72c17f197a9368ac61ff1c31c859699d5bf731a4) can be merged in with the commit that added the initial syntax (9ef7fa87114f70ea9790b9103eb5aa93512d9bcd). 3e1f37007f3ece3957a5e00fac786ae3c0ac2cd1 can also probably be merge in with 9ef7fa87114f70ea9790b9103eb5aa93512d9bcd. You'll have to use some `git rebase -i` magic to make it work but I think it's doable.

**Comment by petroav** [comment]
I had a quick look at your replies. Things look good. I think after you re-shuffle the commits so that everything is cleaner, it would be beneficial if you posted another PR so we can take a look again. Right now I'm having trouble keeping the big picture in my head.

**Comment by amrutagokhale** [comment]
Closing this old ticket.

**Pull request closed without merging**